File: Source/Samples/95_Periodic/code_aselectdropdown.cpp
// ASelectDropdown application source code
#include "PeriodicApp.h"

void PeriodicApp::setup_aselectdropdown( AWidget *layout )
{
    PODVector<AWidget*> dest;
    layout->SearchWidgetClass( "TBButton", dest );
    for (unsigned ii = 0; ii < dest.Size(); ii++)
        SubscribeToEvent(dest[ii], E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleAselectdropdownEvent ));

    AWidget *demo = layout->GetWidget("selectdropdowndemo");
    if ( demo) // warning - this will route for all ASelectDropdown instances events into this event handler.
        demo->SubscribeToEvent(E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleAllSelectdropdownEvent ));
}

void PeriodicApp::HandleAllSelectdropdownEvent(StringHash eventType, VariantMap& eventData)
{
    using namespace WidgetEvent;
    AWidget* widget = static_cast<AWidget*>(eventData[P_TARGET].GetPtr());
    if ( widget == NULL ) return;

    AWidget *demo = widget->FindWidget("selectdropdowndemo");  // find our specific widget
    if ( widget != demo ) return; // if its not ours, bail

    if (eventData[P_TYPE] == UI_EVENT_TYPE_CHANGED )
    {
        if (widget->GetId() ==  "selectdropdowndemo" )
        {
            AppLog( "ASelectDropdown event : " + widget->GetId() + " changed value to " + widget->GetText());
        }
    }
}

void PeriodicApp::HandleAselectdropdownEvent(StringHash eventType, VariantMap& eventData)
{
    using namespace WidgetEvent;
    AWidget* widget = static_cast<AWidget*>(eventData[P_TARGET].GetPtr());
    if ( widget == NULL ) return;
    if (eventData[P_TYPE] == UI_EVENT_TYPE_CLICK)
    {
        if (widget->GetId() == "uiselectdropdowncode" )
        {
            AppLog( "ASelectDropdown support : " + widget->GetId() + " was pressed " );
            ViewCode ( "Sources/code_aselectdropdown.cpp", widget->GetParent() );
        }
        if (widget->GetId() ==  "uiselectdropdownlayout" )
        {
            AppLog( "ASelectDropdown support : " + widget->GetId() + " was pressed ");
            ViewCode ( "Scenes/layout_aselectdropdown.ui.txt", widget->GetParent() );
        }

    }
}

File: bin/Data/AUI/Periodic/LuaScripts/code_aselectdropdown.lua
-- ASelectDropdown application source code
-- when menu entry is selected, the ASelectDropdown value (and string) is updated, theoritically:

function setup_uiselectdropdown( layout )
    local dest = layout:SearchAWidgetClass( "TBButton" )
    for i, awid in ipairs(dest) do
        SubscribeToEvent(awid, "WidgetEvent", "HandleUiselectdropdownEvent")
	end
 
    local demo = layout:GetWidget("selectdropdowndemo")
    if demo ~= nil then
		SubscribeToEvent(demo, "WidgetEvent", "HandleAllSelectdropdownEvent" )
    end
end

function HandleAllSelectdropdownEvent(eventType, eventData)
    local widget = eventData["Target"]:GetPtr("AWidget")
    if widget == nil then
 		return
	end
  	if eventData["Type"]:GetInt() == UI_EVENT_TYPE_CHANGED then   
		AppLog( "ASelectDropdown changed event : " .. widget:GetId() .. " changed value to " .. widget:GetText())
	elseif eventData["Type"]:GetInt() > 0 then
		AppLog( "ASelectDropdown " .. widget:GetId() .. " got event : " .. eventData["Type"]:GetInt())
    end
end

function HandleUiselectdropdownEvent( eventType, eventData)
    local widget = eventData["Target"]:GetPtr("AWidget")
    if widget == nil then
 		return
	end
    if eventData["Type"]:GetInt() == UI_EVENT_TYPE_CLICK then
        if widget:GetId() == "uiselectdropdowncode" then
            AppLog( "ASelectDropdown support : " .. widget:GetId() .. " was pressed " .. widget:GetText() )
            ViewCode ( "LuaScripts/code_aselectdropdown.lua", widget:GetParent())
       end
        if widget:GetId() ==  "uiselectdropdownlayout" then
            AppLog( "ASelectDropdown support : " .. widget:GetId() .. " was pressed ")
            ViewCode ( "Scenes/layout_aselectdropdown.ui.txt", widget:GetParent())
        end
    end
end

